feat: Add Offline DeepSeek Model

varshith257 opened the pull request:
This PR implements an offline `DeepSeek model` loader and inference wrapper fulfilling all requirements in issue. It provides a lightweight, memory-efficient, dependency-minimal way to load and run DeepSeek models from HuggingFace which supports all official DeepSeek R1 & Distill variants using HuggingFace + safetensors. Dynamic model config parsing and deeper memory optimizations (e.g., Triton/offloading) can be addressed in a follow-up issue if needed.

> NOTES TO REVIEWERS

Tested on `DeepSeek-R1-Distill-Qwen-7B` that works offline and follows low-level-only policy and here's o/p in local testing :

```

> python -m unittest intelli.test.integration.test_deepseek_wrapper

----------------------------------------------------------------------
Ran 1 test in 129.257s

OK
Downloading model.safetensors.index.json from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Index downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/91[6](https://github.com/varshith257/Intelli/pull/1/checks#step:6:7)b56a44061fd5cd7d6a8fb63255[7](https://github.com/varshith257/Intelli/pull/1/checks#step:6:8)ed4f724f60/model.safetensors.index.json
Downloading config.json from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Config downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a[8](https://github.com/varshith257/Intelli/pull/1/checks#step:6:9)fb632557ed4f724f60/config.json
Downloading model.safetensors.index.json from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Index downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/[9](https://github.com/varshith257/Intelli/pull/1/checks#step:6:10)16b56a44061fd5cd7d6a8fb632557ed4f724f60/model.safetensors.index.json
Downloading model-00002-of-000002.safetensors from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Model downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a8fb632557ed4f724f60/model-00002-of-000002.safetensors
Downloading model-00001-of-000002.safetensors from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Model downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a8fb632557ed4f724f60/model-00001-of-000002.safetensors
Model weights loaded successfully from split safetensors.
Inference successful, output shape: torch.Size([1, 16, 152064])
```


DETAILED SETUP AND TESTING IS DOCUMENTED IN [README](https://github.com/intelligentnode/Intelli/pull/119/files#diff-54ae69beee243b0dd0e9d51d72adefed7108c31b8d18d9d90ccca13c97c0bb66)

/claim #82

varshith257 commented:
@intelligentnode Any reviews on this :)

intelligentnode commented:
Thanks for the attempt, it is good you used only low level dependencies.

Kindly answer the following questions to help with the review:
- There is no tokenizer implementation. How will text be converted to token IDs ?
- While the code tested using `Qwen-7B`, can the same code run deepseek R1 ?
- Does this implementation work on both CPU and GPU devices ? How to test in GPU and which device to use ?

varshith257 commented:
> Thanks for the attempt, it is good you used only low level dependencies.
> 
> Kindly answer the following questions to help with the review:
> 
> * There is no tokenizer implementation. How will text be converted to token IDs ?
> * While the code tested using `Qwen-7B`, can the same code run deepseek R1 ?

Yes, the implementation is architecture-agnostic, we rely on `config.json` + `safetensors.index.json` it can load any DeepSeek-R1 variant including the full 671B model when provided sufficient RAM/VRAM

> * Does this implementation work on both CPU and GPU devices ? How to test in GPU and which device to use ?

Yes it works both with the GPU and CPU and to test with GPU you can use
```
CUDA_VISIBLE_DEVICES=0 python -m unittest intelli.test.integration.test_deepseek_wrapper
```
For full DeepSeek-R1, the recommend testing could be on A100 80GB or A6000 48GB class GPUs. Quantized versions can fit in 35–40GB VRAM and are loadable via our implementation using `safetensors.index.json` split loading

varshith257 commented:
Let me know anything to add or iterate

Barqawiz commented on intelli/test/integration/test_deepseek_wrapper.py:
```diff
@@ -0,0 +1,30 @@
+
+        self.assertEqual(output.shape[1], dummy_input.shape[1])
+        print("Inference successful, output shape:", output.shape)
+
```
Extend the test case for input text and output text, not only dummy int. And include all components in test such as tokenizer.

varshith257 replied:
Now, I have tested the full pipeline and here's o/p of tests ran:

```
----------------------------------------------------------------------
Ran 2 tests in 202.735s
OK
Downloading model.safetensors.index.json from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Index downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/91[6](https://github.com/varshith257/Intelli/actions/runs/14565721476/job/40854657398#step:6:7)b56a44061fd5cd7d6a8fb63255[7](https://github.com/varshith257/Intelli/actions/runs/14565721476/job/40854657398#step:6:8)ed4f724f60/model.safetensors.index.json
Downloading config.json from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Config downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a[8](https://github.com/varshith257/Intelli/actions/runs/14565721476/job/40854657398#step:6:9)fb632557ed4f724f60/config.json
Downloading model.safetensors.index.json from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Index downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/[9](https://github.com/varshith257/Intelli/actions/runs/14565721476/job/40854657398#step:6:10)16b56a44061fd5cd7d6a8fb632557ed4f724f60/model.safetensors.index.json
Downloading model-00001-of-000002.safetensors from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Model downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a8fb632557ed4f724f60/model-00001-of-000002.safetensors
Downloading model-00002-of-000002.safetensors from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Model downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a8fb632557ed4f724f60/model-00002-of-000002.safetensors
Model weights loaded successfully from split safetensors.
Inference successful, output shape: torch.Size([1, 16, 152064])
Downloading model.safetensors.index.json from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Index downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a8fb632557ed4f724f60/model.safetensors.index.json
Downloading config.json from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Config downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a8fb632557ed4f724f60/config.json
Downloading model.safetensors.index.json from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Index downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a8fb632557ed4f724f60/model.safetensors.index.json
Downloading model-00001-of-000002.safetensors from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Model downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a8fb632557ed4f724f60/model-00001-of-000002.safetensors
Downloading model-00002-of-000002.safetensors from deepseek-ai/DeepSeek-R1-Distill-Qwen-7B...
Model downloaded to /home/runner/.cache/deepseek/models--deepseek-ai--DeepSeek-R1-Distill-Qwen-7B/snapshots/916b56a44061fd5cd7d6a8fb632557ed4f724f60/model-00002-of-000002.safetensors
Model weights loaded successfully from split safetensors.
Token IDs: [1986, 285, 64, [12](https://github.com/varshith257/Intelli/actions/runs/14565721476/job/40854657398#step:6:13)8244, 128244]
Inference successful on text input, output shape: torch.Size([1, 5, 152064])

```

Barqawiz commented on requirements.txt:
```diff
@@ -1,4 +1,9 @@
+safetensors>=0.5.0
+triton>=2.1.0
+huggingface-hub>=0.30.0
+tokenizers>=0.20.0
```
I don't want to use tokenizers module. 
It is considered high level.

varshith257 replied:
Done! Added a basic helper for tokenization

intelligentnode commented:
Deepseek use Byte-level BPE (BBPE), not whitespace word tokenizer.

Please adjust.

intelligentnode commented:
Feedback:
- There are no test cases to validate the output generate valid output. add test cases to generate test and validate the string readable.
- When I switch the model name to "deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B" and error printed. The code should support all deepseek models.
_Entry Not Found for url: https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B/resolve/main/model.safetensors.index.json._
-

varshith257 commented:
@intelligentnode Thanks for detailed review and yes some variants aren't using shared indices, I have fixed them now and as we discussed earlier i implemented the basic BPE tokenizer and also implemented decode fn to test make it more redable now


And testing command has been updated. Kindly use thie following cd to test them now it solves dynamically 

```
DEEPSEEK_MODEL= <model_name> python -m unittest intelli.test.integration.test_deepseek_wrapper
```

varshith257 commented:
@intelligentnode @Barqawiz Can you review this, please?

intelligentnode commented:
There is conflicts and the outputs not working as expected.

varshith257 commented:
@intelligentnode Can you paste the logs you qre getting? I tested it and got fully decode roundtrip string

varshith257 commented:
Hey @intelligentnode, the issue was closed citing time-sensitivity, but no clear deadline was mentioned upfront. The bounty was also reduced mid-way . Contributors put in genuine effort based on the original scope and that should be acknowledged and fairly rewarded. Could we get some clarity on this?

intelligentnode commented:
The issue is that there was no testing conducted during this contribution, and there seemed to be limited understanding of DeepSeek’s capabilities. The process of testing and providing feedback became quite lengthy. Since DeepSeek is a time-sensitive topic and the models evolve quickly, this delay impacted our ability to proceed efficiently.
Look to the history I gave fair chance for this change. Please also note that the bounty platform does not support setting explicit deadlines.
